allow environment to be passed to the HTML page

bogdanRada opened the pull request:
Hello, 
  I made few changes to allow the environment to be passed from the rack middleware to the HTML page. This is especially needed when you have other middlewares in your middleware stack that defines some variables in the environment that needed to be passed to the HTML page. 

Please let me know what you think. Thank you very much.

adamcrown commented:
Hi @bogdanRada. Thanks for the pull request. The change sounds reasonable, but I guess I'm confused about how you're going to actually implement it.

Turnout doesn't support ERB in the HTML templates. Currently it's a Liquid style syntax, but even that isn't really Liquid since that seemed like overkill for just the `{{ reason }}` tag. So it's just a simple `gsub` to get `@reason` into the HTML template right now.

So how exactly are you going to use `@rack_env`  in your page?

Thanks for clarifying this for me.

bogdanRada commented:
i have multiple rack middlewares in my middleware stack and one of the middlewares is detecting the brand of the application and sets some configurations based on that. The brand howewer is set as a environment variable by this rack middleware and i can not access it unless i do some hacking. And i need to have access to this environment variable because i also need to display not only different reasons for maintenance ( that contain branded specific information)  but also different assets based on the brand. 

This is currently the hack i am using and this pull request would help me get rid of the hack. I really don't like to monkey patch gems , because it will be harder in future to maintain the patches. 

```ruby
Rack::Turnout.class_eval do
  alias_method :call_orig, :call

  # we need to rewrite the call method and save our brand
  def call(env)
    MyAwesomeMiddleware.new(@app)

    request = Turnout::Request.new(env)
    settings = Turnout::MaintenanceFile.find

    if settings && !request.allowed?(settings)
      page_class = Turnout::MaintenancePage.best_for(env)
      page = page_class.new(settings.reason)
      page.instance_eval { @brand = env['BRAND'] }

      page.rack_response(settings.response_code, settings.retry_after)
    else
      @app.call(env)
    end
  end
end
````
And currently i am using Nokogiri to parse the content and insert the correct information like this

```ruby
Turnout::MaintenancePage::HTML.class_eval do

  #adding html content to our default page
  def content
    brand = Rails.configuration.brands[@brand]
    content = File.open(Rails.configuration.maintenance_page, 'rb').read

    html = Nokogiri::HTML(content)
     ........
     # using here the html instance class to insert the information
     # code has been stripped from here for confidential reasons
     ........
    content = html.to_s

    content
  end
end

```
And if you accept this pull request, it will probably help me get rid of both monkey patches. 

If you have other questions , please let me know.

adamcrown commented:
Thanks for the thorough explanation. The problem I see is that defining `@rack_env` without any way to use it without monkey patching, wouldn't be helpful to the average user without monkey patching.

I will, however, consider adding support for `maintenance.html.erb` files so that the environment hash can be used in the maintenance files. If you'd like to create a pull request for that, that would be great. Otherwise I'll see if I can work on it when I have some downtime.

bogdanRada commented:
will definitely try my best to do those changes :) thank you

bogdanRada commented:
i added Tilt gem (https://github.com/rtomayko/tilt) as dependency to turnout for rendering templates.
It supports multiple engines like ERB, HAML, Liquid, so if anyone uses liquid template for their maintenance pages, they should not see any difference. 

Also made the class Turnout::MaintenancePage::Base to acccept an option list , in case something else will be needed in future.  

Let me know what you think. Thank you very much 

Travis CI build failing seems to be not related to the changes: i think this is the reason: https://github.com/rubygems/rubygems/issues/1220

bogdanRada commented:
Still working on this though . Would be nice if i could find a way to use ActionView helpers in templates . 
I am trying to research that.

adamcrown commented:
Thanks for working on this. Adding Tilt sounds fine but I'm a bit hesitant to add actionview as a dependency since it has so many other dependencies and I think that would add a bit too much bloat for non-Ralis apps.

bogdanRada commented:
i already did the changes for actionview here: https://github.com/bogdanRada/turnout/tree/feature/actionview . And they do seem to work. Just need to test it more .

bogdanRada commented:
i have fixed the tests, and i have moved the Tilt rendering into its own class and added a check if File exists in the lib/turnout/maintenance_page.rb . Let me know what you think if you still want to proceed with Tilt or use Actionview. Thank you very much

adamcrown commented:
This looks pretty good. I'd like to try this out and test the code myself before merging it but I'm a bit busy right now. I'll do my best to get to it soon though.

Thanks for all the work on this.

bogdanRada commented:
Great :D Thank you very much

bogdanRada commented:
Just one thing. I think we could relax a bit the version requirement for tilt to be >= 1.4 and < 2, because it seems actionpack and sprockets depend on tilt and Rails version that are >= 4.0 are also dependent to tilt 1.4 . what do you think?

bogdanRada commented:
any news on this? Does this PR need anything else to be merged? Or any other changes? I am opened to suggestions .

adamcrown commented:
I apologize it's taken me so long. I wanted to play around with it before merging it because it changes so much. I've been pretty busy and I have a project deadline coming up now. But I'll make it a priority to get to it the next time I've got some free time.

bogdanRada commented:
ok, thanks. Let me know if there is anything i can do to help you.

adamcrown commented on lib/turnout/maintenance_page.rb:
```diff
@@ -11,14 +12,14 @@ def self.best_for(env)
       best_type = request.best_media_type(all_types)
-
-      best = all.find { |page| page.media_types.include? best_type }
+      best = all.find { |page| page.media_types.include?(best_type) && File.exist?(page.new.path) }
```
The `File.exist?` call is always returning `true` because the paths always exist in the gem. I think we'll have to first check if the custom path exists and if none of those are found then go with the default page in the gem. Does that make sense? If you have time to make this change, that would be great. Otherwise I'll see if I can get to it.

bogdanRada commented:
Thanks for the comments. Will take a look hopefully tomorrow and see how can i improve that.

bogdanRada commented:
i have fixed the issue you mentioned. Let me know if it is better now or something else needs to be changed or improved. Thank you very much

adamcrown commented:
Finally got a chance to try out the latest version. It looks good. Thanks for your patience and all your work on this. I'll update the README and push out a new version soon.

bogdanRada commented:
awesome :smile:

adamcrown commented:
So I do hate to delay this any further but I had a realization as I was writing the README. There are a couple of unlikely but possible security concerns with having `.erb` support.

- The file is in the public directory so it's source code will be readable by visiting `example.com/maintenance.html.erb`.
- If uploads are allowed directly in the `/public` directory, then it would allow any code to be executed. 

The second one especyally shouldn't every happen because the `/public/assets` directory also needs to be protected from user uploads. But it's still something to keep in mind. And I doubt anybody would put any sensitive information in their maintenance page file. But I do want to be extra cautious about security.

So my thought is to make a small change to disable the ERB pages by default with the option to enable them for the people who need it. I'll then add some warnings to the README. Eventually I may just move to a totally different location outside of `/public` to store the custom maintenance page templates.

How does that sound to you?

bogdanRada commented:
Hello,
  I do understand the first part that the souce code might be readable if you visit example.com/maintenance.html.erb although this can be prevented by htaccess or virtual host configuration and also using a robots.txt to prevent spiders trying to read the content. 
  And it does make sense that uploads should not be allowed to override the maintenance files. But this is something that could happen with older versions as well, so that should be taken in consideration with any version, regardless if erb support is enabled or not . but i think any good developer knows that uploads should not override or rewrite application files.
 And about displaying sensitive information on maintenance pages , i really doubt this would ever happen, because if maintenance mode is enabled, this pages would however be visible to anyone. So i think we should not be concerned about this aspect. 

   I do like the idea though of moving the ERB template to other directory. I have been thinking about this too.  Tilt is capable to render it from any location, so people that use Rails might want to put it into **app/views** directory. 
   I think if we just move the ERB template to another directory everything would be ok, because no one could access it directly from browser anymore, so first point would be solved.
 You can put in the README into the TIPS section the information about not allowing uploads in the same directory, but i would not consider this as a important security concern. Because anyone can customize their own path to their own maintenance pages by setting the **app_root** when instantiating the Rack midldeware. So i really think this is out of this gem's control and it's up to the user to decide where he wants this files to reside .

adamcrown commented:
I had some time to work on this. I added a `maintenance_pages_path` setting and a warning to the README. I feel this covers the bases sufficiently. So I released version 2.3.0. I hope this takes care of your needs. Thanks for all of your work on this. It's a nice new feature to have.

bogdanRada commented:
thank you also :)
